src/librustc_mir/interpret/operator.rs

```diff
@@ -197,94 +137,129 @@ impl<'a, 'mir, 'tcx, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M> {
             }
         };
 
-        let val = match (bin_op, left_kind) {
-            (_, F32) => float_op(bin_op, l, r, FloatTy::F32),
-            (_, F64) => float_op(bin_op, l, r, FloatTy::F64),
-
-
-            (Eq, _) => PrimVal::from_bool(l == r),
-            (Ne, _) => PrimVal::from_bool(l != r),
-
-            (Lt, k) if k.is_signed_int() => PrimVal::from_bool((l as i128) < (r as i128)),
-            (Lt, _) => PrimVal::from_bool(l < r),
-            (Le, k) if k.is_signed_int() => PrimVal::from_bool((l as i128) <= (r as i128)),
-            (Le, _) => PrimVal::from_bool(l <= r),
-            (Gt, k) if k.is_signed_int() => PrimVal::from_bool((l as i128) > (r as i128)),
-            (Gt, _) => PrimVal::from_bool(l > r),
-            (Ge, k) if k.is_signed_int() => PrimVal::from_bool((l as i128) >= (r as i128)),
-            (Ge, _) => PrimVal::from_bool(l >= r),
+        if left_ty.is_signed() {
+            let op: Option<fn(&i128, &i128) -> bool> = match bin_op {
+                Lt => Some(i128::lt),
+                Le => Some(i128::le),
+                Gt => Some(i128::gt),
+                Ge => Some(i128::ge),
+                _ => None,
+            };
+            if let Some(op) = op {
+                let l = self.sign_extend(l, left_ty)? as i128;
+                let r = self.sign_extend(r, right_ty)? as i128;
+                return Ok((PrimVal::from_bool(op(&l, &r)), false));
+            }
+            let op: Option<fn(i128, i128) -> (i128, bool)> = match bin_op {
+                Rem | Div if r == 0 => return Ok((PrimVal::Bytes(l), true)),
+                Div => Some(i128::overflowing_div),
+                Rem => Some(i128::overflowing_rem),
+                Add => Some(i128::overflowing_add),
+                Sub => Some(i128::overflowing_sub),
+                Mul => Some(i128::overflowing_mul),
+                _ => None,
+            };
+            if let Some(op) = op {
+                let l128 = self.sign_extend(l, left_ty)? as i128;
+                let r = self.sign_extend(r, right_ty)? as i128;
+                let size = self.layout_of(left_ty)?.size.bits();
+                match bin_op {
+                    Rem | Div => {
+                        // int_min / -1
+                        if r == -1 && l == (1 << (size - 1)) {
+                            return Ok((PrimVal::Bytes(l), true));
+                        }
+                    },
+                    _ => {},
+                }
+                trace!("{}, {}, {}", l, l128, r);
+                let (result, mut oflo) = op(l128, r);
+                trace!("{}, {}", result, oflo);
+                if !oflo && size != 128 {
+                    let max = 1 << (size - 1);
+                    oflo = result >= max || result < -max;
```

Comment:
Can't you write this as `sign_extend(truncated, left_ty) != truncated`?

Reply:
I think this is much clearer. Also we'd run into the difference of the zeroes and ones on the left side again